keps/sig-scheduling/3990-pod-topology-spread-fallback-mode/README.md

```diff
@@ -0,0 +1,1023 @@
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+- 2023-08-12: Initial KEP PR is submitted.
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### introduce `DoNotScheduleUntilScaleUpFailed` and `DoNotScheduleUntilPreemptionFailed`
+
+Instead of `FallBackCriteria`, introduce `DoNotScheduleUntilScaleUpFailed` and `DoNotScheduleUntilPreemptionFailed` in `WhenUnsatisfiable`.
+`DoNotScheduleUntilScaleUpFailed` corresponds to `ScaleUpFailed`, 
+and `DoNotScheduleUntilPreemptionFailed` corresponds to `PreemptionFailed`.
+
+We noticed a downside in this way, compared to `FallBackCriteria`.
+In other scheduling constraints, we distinguish between preferred and required constraint by where the constraint is written in.
+For example, PodAffinity and NodeAffinity, if it's written in `requiredDuringSchedulingIgnoredDuringExecution`, it's required. 
+And if it's written in `preferredDuringSchedulingIgnoredDuringExecution`, it's preferred.
+
+In the future, we may want to introduce similar fallback mechanism in such other scheduling constraints, 
+but, we couldn't make the similar API design if we went with `DoNotScheduleUntilScaleUpFailed` and `DoNotScheduleUntilPreemptionFailed`, 
+as they don't define preferred or required in enum value like `WhenUnsatisfiable`.
+
+On the other hand, `FallBackCriteria` allows us to unify APIs in all scheduling constraints. 
+We will just introduce `FallBackCriteria` field in them and there we go.
```

Comment:
This makes me think that we might want to model fallback criteria at the pod level. 
```
preferentialFallback:
  - PodAffinity
  - AntiAffinity
  - TopologySpreadConstraints
```
It may be difficult to get the granularity right, though. At another extreme, you'd add a preference policy for each preferred term across the pod spec.

Reply:
That's an interesting idea. 
Because when people want to use fallback, they may want to do the same in all required scheduling constraints the Pod has. Also, we may want to give `preferentialFallback` the way to "fallback in all".

So, maybe it'd be like this:

```yaml
preferentialFallback:
  schedulingconstraints: 
  - "*" # you can specify individual scheduling constraints name (PodAffinity, NodeAffinity, etc) as well.
  fallBackCriteria:
  - ScaleUpFailed
```

Let me include this idea in the alternative section for now. Like the idea of introducing enums, we can gather feedback from more people about the design choice among them.